### PR TITLE
Allow text strings to special type to date [ci drivers]

### LINKED ISSUE #1576: Add Special Type: DateTime String (ISO)
UNIX Timestamp is very handy but it would be awesome to have a few formats supporting datetime strings:  Date, Time, DateTime.  I'm sure there are lots of silly date and time formatting options.  I imagine that support ISO formats only would be a good first step.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by dpsutton**
Fixes #1576 

Ability to have text columns used as dates

Caveats:
- most databases blow up if the text string is not coercible. Some just return a null (mysql and variants).
- not all dbs are supported. All databases that go through `metabase.driver.sql-jdbc` has a default and all codepaths through `metabase.driver.sql.query-processor` has this called. Some databases provide their own query processor (bigquery, googleanalytics, druid, mongo) and these do not support this. These largely tracks the support for `:type/UNIXTimestampMilliseconds` and seconds. Mongo independently implements unix timestamps in the query processor but presumably the others are completely silent on this feature. (The unix timestamps feature is implemented in the same sql.query-processor as the feature under review here and unix timestamps are implemented independently in the mongo query processor).
- There is no feedback mechanism for when this feature is not supported. Similar to `:type/UNIXTimestampMilleseconds` on unsupported databases, not every db supports this feature. At the moment there is no mechanism to provide feedback for this. Its possible that the list of special_types is filtered to remove this, or perhaps some validation when selecting, or perhaps make the query processor aware of supported special_types and throw a proper error message. As it stands now, it will be silently ignored on databases that do not support it.
- Some databases are only partially supported. Oracle (as far as I can tell) has no time datatype. This means that the date and datetime variants are supported but it lacks a way to express `:type/ISOTimeString`. Attempting to use it will throw an invalid cast in the query processor. It would be easy to add 
```clojure
(defmethod sql.qp/cast-temporal-string [:oracle :type/ISO8601TimeString]
  [_driver _special_type expr]
  (throw (Exception. (tru "Oracle does not support ISO8601TimeString"))))
```
and perhaps I should.
- the fields aren't queryable as dates on oracle and sparksql. I haven't investigated this in depth yet but was a bit surprised. Oracle throws a NPE in the driver and sparksql throws a parse exception `DataType time() is not supported.(line 2, pos 178)`.
```clojure
(testing "are queryable as dates"
  (testing "a datetime field"
    ;; TODO: why does this fail on oracle? gives a NPE
    (mt/test-drivers (disj (sql-jdbc.tu/sql-jdbc-drivers) :oracle :sparksql)
      (is (= 1
             (count (mt/rows (mt/dataset string-times
                               (mt/run-mbql-query times
                                 {:filter   [:= [:datetime-field $ts :day] "2008-10-19"]}))))))))
  (testing "a date field"
    (mt/test-drivers (disj (sql-jdbc.tu/sql-jdbc-drivers) :oracle :sparksql)
      (is (= 1
             (count (mt/rows (mt/dataset string-times
                               (mt/run-mbql-query times
                                 {:filter   [:= [:datetime-field $d :day] "2008-10-19"]})))))))))
```

# For reviewers:

Changes should be relatively straight forward. Things I'm particularly interested in feedback on:
- should the defmethods and default implementations be in better places to give more database drivers an opportunity to hook into them? Ie, `cast-temporal-string` being in `metabase.driver.sql.query-processor` and the defaults being in `metabase.driver.sql-jdbc`.
- are any of the above caveats necessary for the first release? Do any merit follow up tickets now?

### Things to note
For future development we can use `isa?` in match expressions.
```clojure
(defn cast-field-if-needed
  "Wrap a `field-identifier` in appropriate HoneySQL expressions if it refers to a UNIX timestamp Field."
  [driver field field-identifier]
  (match [(:base_type field) (:special_type field)]
    [_ (:isa? :type/UNIXTimestampSeconds)]      (unix-timestamp->honeysql driver :seconds      field-identifier)
    [_ (:isa? :type/UNIXTimestampMilliseconds)] (unix-timestamp->honeysql driver :milliseconds field-identifier)
    [:type/Text (:isa? :type/TemporalString)]   (cast-temporal-string driver (:special_type field) field-identifier)
    :else field-identifier))
```

Very simple implementation:
```clojure
(defmethod match/emit-pattern-for-syntax [:isa? :default]
  [[_ parent]] {::match/tag ::isa? :parent parent})

(defmethod match/to-source ::isa?
  [{parent :parent} ocr]
  `(isa? ~ocr ~parent))
```

In the future, it would be simple to add the ability to add a hierarchy into this match. If its desired to add a way to capture the type, this would require moving to a custom deftype and implementing the `ISpecializeMatrix` protocol which gets a bit more involved with the matching algorithm. Not super difficult but there's not a ton of examples out there and requires knowledge of how the pattern matrices are made.

### CODE REVIEW

**Comment by sbelak** [comment]
Since presumably this will only come into effect if the user manually overwrites the base type to date, I wouldn't worry about formats and assume uniformity. If either of those are not true they have bigger problems than us not picking up their dates.

**Comment by camsaul** [comment]
> Since presumably this will only come into effect if the user manually overwrites the base type to date

@sbelak If I understand this PR correctly, they'd be leaving the base type as `:type/Text` and changing the special type to `:type/TextDate`. The base type isn't meant to be user-overwritable

File: src/metabase/types.clj [hunk 1]
```diff
@@ -109,6 +109,7 @@
 (derive :type/UNIXTimestamp :type/Integer)
 (derive :type/UNIXTimestampSeconds :type/UNIXTimestamp)
 (derive :type/UNIXTimestampMilliseconds :type/UNIXTimestamp)
+(derive :type/TextDate :type/DateTime)
```
Comment (camsaul) [inline src/metabase/types.clj:112]: a Date is not a Date**Time**

File: frontend/src/metabase/lib/core.js [hunk 2]
```diff
@@ -223,6 +223,11 @@ export const field_special_types = [
     name: t`Updated timestamp`,
     section: t`Date and Time`,
   },
+  {
+    id: TYPE.TextDate,
+    name: t`Text as a timestamp`,
+    section: t`Date and Time`,
```
Comment (camsaul) [inline frontend/src/metabase/lib/core.js:229]: If it's a date and time then don't call it `TextDate`, call it `TextDateTime`. A `TextDate` in my mind would be `2020-11-23`

**Comment by camsaul** [comment]
@dpsutton I don't think just having `:type/TextDate` is going to end up being sufficient. You probably want to do something like have `:type/Text` -> `:type/TemporalString` -> `:type/DateString`/`:type/DateTimeString`/`:type/TimeString` at the very least. I can tell you with 100% certainty that there are some databases out there that are not going to let you use the same function to parse `2020-11-23`, `2020-11-23T17:49:00`, and `17:49:00`.

If you just want to support `:type/TextDateTime` or `:type/DateTimeString` or whatever and add the other two later, at least leave the option to add support for the others

**Comment by camsaul** [comment]
@dpsutton as you mentioned, we probably need different types for different formats as well. Similar to how we have both `UNIXTimestampSeconds` and `UNIXTimestampMilliseconds` -- basically the same idea but two different formats. Is the date it ISO-8601? A standard SQL timestamp like `2004-10-19 10:23:54`? `YYYYMMDD`? For a lot of DBs when parsing temporal strings we need to give it a format string.

It would be nice to be able to let users specify arbitrary format strings but that would be a pretty big change... in the mean time maybe it's better just to offer a few common formats

**Comment by camsaul** [comment]
Related issue: https://github.com/metabase/metabase/issues/1576

**Comment by camsaul** [comment]
Besides the things I noted, this functionality is long overdue and I'm a big 👍 on getting this PR merged. I'm happy if we just shipped ISO-8601 DateTime support for now, but I just want to make sure we're not going to end up with a dozen "my `MM/DD/YYYY`-formatted column doesn't work" issues after we ship this feature -- we should be explicit about what we're going to support and leave the door open to support more stuff in the future.

**Comment by mysticaltech** [comment]
@camsaul Makes sense! ISO-8601 would be a great start. @dpsutton keep up the good work!

**Comment by mysticaltech** [comment]
Keep up the good work @dpsutton! Presto support would be mind-bending.

**Comment by dpsutton** [comment]
@mysticaltech all sql databases should get this by default. I haven't run presto locally with this change but the `alternative-date-tests` run against the presto driver:

```
# output from the presto driver tests
Testing ********.query-processor-test.alternative-date-test
Sync :presto Database sad-toucan-incidents took 10.6 s
Sync :presto Database tupac-sightings took 17.2 s
load-data for h2 string-times times took 5.7 ms # <- this is the db with the string date types
Sync :h2 Database string-times took 266.6 ms
```

It's running against the `string-times` sample database so it seems this is working for presto.

File: src/metabase/driver/sql/query_processor.clj [hunk 3]
```diff
@@ -209,6 +210,16 @@
   (fn [driver seconds-or-milliseconds _] [(driver/dispatch-on-initialized-driver driver) seconds-or-milliseconds])
   :hierarchy #'driver/hierarchy)
 
+(defmulti cast-temporal-string
```
Comment (sbelak) [inline src/metabase/driver/sql/query_processor.clj:213]: I think it's reasonable to have these here. Non-SQL as they are in a lot of ways an artefact of SQL semantics. If we ever come to a situation that we want to support this for other drivers, we can always move it to the main driver namespace, but for now I think smaller surface area wins.

**Review by sbelak — APPROVED** [review]
Concur with Cam. Looks good and a great addition to MB!

File: src/metabase/driver/sql/query_processor.clj (hunk 3 above)
Comment (dpsutton) [reply src/metabase/driver/sql/query_processor.clj:213]: thanks @sbelak . It felt right and kinda mimiced what the unix timestamp was doing but was worried i was foreclosing other databases unnecessarily.

Comment (mysticaltech) [reply src/metabase/driver/sql/query_processor.clj:213]: Well done! 🏆

**Review by camsaul — APPROVED** [review]
Awesome

**Pull request merged**